rcl/src/rcl/logging_rosout.c

```diff
@@ -0,0 +1,236 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__LOGGING_H_
+#define RCL__LOGGING_H_
+
+
+#include "rcl/allocator.h"
+#include "rcl/error_handling.h"
+#include "rcl/node.h"
+#include "rcl/publisher.h"
+#include "rcl/types.h"
+#include "rcl/visibility_control.h"
+#include "rcl_interfaces/msg/log.h"
+#include "rcutils/allocator.h"
+#include "rcutils/macros.h"
+#include "rcutils/types/hash_map.h"
+#include "rosidl_generator_c/string_functions.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#define ROSOUT_TOPIC_NAME "rosout"
+
+/* Return RCL_RET_OK from this macro because we won't check throughout rcl if rosout is
+ * initialized or not and in the case it's not we want things to continue working.
+ */
+#define RCL_LOGGING_ROSOUT_VERIFY_INITIALIZED \
+  if (!__is_initialized) { \
+    return RCL_RET_OK; \
+  }
+
+typedef struct rosout_map_entry_t
+{
+  rcl_node_t * node;
+  rcl_publisher_t publisher;
+} rosout_map_entry_t;
+
+static rcutils_hash_map_t __logger_map;
+static bool __is_initialized = false;
+static rcl_allocator_t __rosout_allocator;
+
+rcl_ret_t rcl_logging_rosout_init(
+  const rcl_allocator_t * allocator)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(allocator, RCL_RET_INVALID_ARGUMENT);
+  rcl_ret_t status = RCL_RET_OK;
+  if (__is_initialized) {
+    return RCL_RET_OK;
```

Comment:
I think I had to change this in order to fix unit tests because they were causing initialize to be called multiple times. I've corrected the documentation in the header to match.